src/core/dev/allocator.cpp

```diff
@@ -363,3 +379,195 @@ bool xlio_allocator_hw::register_memory(ib_ctx_handler *p_ib_ctx_h)
 {
     return m_data && xlio_registrator::register_memory(m_data, m_size, p_ib_ctx_h);
 }
+
+/*
+ * xlio_allocator_heap implementation
+ */
+
+struct heap_key {
+    alloc_t alloc_func;
+    free_t free_func;
+    bool hw;
+
+    bool operator==(const heap_key &key) const
+    {
+        return key.alloc_func == alloc_func && key.free_func == free_func && key.hw == hw;
+    }
+};
+
+namespace std {
+template <> class hash<heap_key> {
+public:
+    size_t operator()(const heap_key &key) const
+    {
+        return (((size_t)key.alloc_func + (size_t)key.free_func) & ~(size_t)1LU) | key.hw;
```

Comment:
maybe will be faster to use XORs instead of addition and NOT+OR:
return ((size_t)key.alloc_func ^ (size_t)key.free_func ^ key.hw);

Reply:
Good point, will change it